### PR TITLE
Parser: Runs all parser implementations against the same tests

### PR DESCRIPTION
**Pull request by dmsnell**
So far we haven't added too many tests to the parsers but the ones we
_have_ added are in separate places and each implementation runs against
a different set of tests.

In this patch we're creating `shared-tests.js` in the spec parser that
defines a suite of conformance tests for the parser and then we're using
that base test-builder to dynamically create test suites for each
implementation such that they all run the same suite.

It's probably easier to understand by reading the code than this
summary.

Of note: by calling to `php` directly we're able to run the PHP parsers
against the same tests as we are the JavaScript implementations. We
should be able to do this for any implementation as long as the required
binaries are available in the CI environment (Rust, for example).

**Update** Now this only runs the PHP tests if `php` is available and running.
The CI environment runs `php` fine and so does a local build if `php` is installed.
There's no major need to have the PHP parser tests run on every build though and
in order to not block people who are working on JS these tests will now skip in
such a case.

**PHP not available**
screen shot 2018-10-31 at 2 41 40 pm

**PHP available**
screen shot 2018-10-31 at 2 42 00 pm

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation. (? maybe)

### CODE REVIEW

File: packages/block-serialization-default-parser/test/index.js [hunk 1]
```diff
@@ -1,73 +1,24 @@
...
-			'<!-- wp:block /--><p>Break me</p>',
-			'<!-- wp:block --><!-- /wp:block --><p>Break me</p>',
-		].forEach( ( input ) => expect( parse( input ) ).toEqual( [
-			expect.objectContaining( { blockName: 'core/block', innerHTML: '' } ),
-			expect.objectContaining( { innerHTML: '<p>Break me</p>' } ),
-		] ) );
-	} );
-} );
+describe( 'block-serialization-default-parser-js', testParser( parse ) );
+describe( 'block-serialization-default-parser-php', testParser( ( document ) => JSON.parse(
```
Comment (aduth) [inline packages/block-serialization-default-parser/test/index.js]: I expect this may be shadowing the global DOM `document`.

**Review by aduth — APPROVED** [review]
Would be good if we can be tolerant to environments where PHP is not available, though not strictly a blocker.

File: packages/block-serialization-default-parser/test/index.js (hunk 1 above)
Comment (dmsnell) [reply packages/block-serialization-default-parser/test/index.js]: it will be. I forgot we are avoiding that…will change

**Comment by dmsnell** [comment]
Mobile tests are failing because of an error with the deprecated `setUnknownTypeHandlerName()` function and they don't appear to be related to this PR. I'll go ahead and merge with those failing tests as long as the other tests pass.

Tests are running all clear locally for me.

File: packages/block-serialization-spec-parser/test/__snapshots__/index.js.snap [hunk 2]
```diff
@@ -10,3 +10,14 @@ Array [
   },
 ]
 `;
+
+exports[`block-serialization-spec-parser-php basic parsing parse() works properly 1`] = `
+Array [
+  Object {
+    "attrs": Array [],
```
Comment (gziolo) [inline packages/block-serialization-spec-parser/test/__snapshots__/index.js.snap]: This is concerning that it produces different type for `attrs` property.

Should we add test which validates that both outputs generated by parsers are deeply equal?

Comment (gziolo) [reply packages/block-serialization-spec-parser/test/__snapshots__/index.js.snap]: I'm just leaving a note that it doesn't work as expected. The second parser doesn't have this issue  :)

**Review by gziolo — APPROVED** [review]
Nice, I like it. I got scared at first seeing PHP file included, but it makes perfect sense to consolidate it here to be able to share tests. Awesome idea 👍

File: packages/block-serialization-default-parser/test/index.js [hunk 3]
```diff
@@ -1,73 +1,14 @@
+/**
+ * External dependencies
+ */
+import path from 'path';
+
 /**
  * Internal dependencies
  */
+import { jsTester, phpTester } from '../../block-serialization-spec-parser/shared-tests';
```
[the remark this answers is not in the document]
Comment (dmsnell) [reply packages/block-serialization-default-parser/test/index.js]: I changed to use this and then moved `jsTester` and `phpTester` into `index.js`. this created a new problem which was the import of `node` modules in the code.

to work around this I hid `require( 'child_process' )` behind a `'test' === process.env.NODE_ENV` check

do you think this is a reasonable mechanism to keep the unwanted `node` code out of the runtime while preserving it in the tests?

**Pull request merged**